packages/router/router.ts

```diff
@@ -2319,26 +2324,40 @@ export function createRouter(init: RouterInit): Router {
       replace?: boolean;
     } = {}
   ) {
-    if (redirect.revalidate) {
+    if (redirect.response.headers.has("X-Remix-Revalidate")) {
       isRevalidationRequired = true;
     }
 
-    let redirectLocation = createLocation(state.location, redirect.location, {
+    let location = redirect.response.headers.get("Location");
+    invariant(location, "Expected a Location header on the redirect Response");
+    let redirectLocation = createLocation(state.location, location, {
       _isRedirect: true,
     });
-    invariant(
-      redirectLocation,
-      "Expected a location on the redirect navigation"
-    );
+
+    if (ABSOLUTE_URL_REGEX.test(location)) {
```

Comment:
This redirect logic was moved out of `callLoaderOrAction` to genericize it for use with `dataStrategy`